lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,53 +39,147 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(text, str) or not isinstance(end_of_word, str) or \
+            not (isinstance(start_of_word, str) or start_of_word is None):
+        return None
+
+    words = text.split()
+    prepared_words = [prepare_word(word, start_of_word, end_of_word) for word in words]
+
+    freq = {}
+    for word in set(prepared_words):
+        if word is None:
+            return None
+        freq[word] = prepared_words.count(word)
+
+    return freq
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    seq_freq = {}
+    for key in word_frequencies:
+        for token_index in range(len(key) - 1):
+            tokens = (key[token_index], key[token_index + 1])
+            if tokens not in seq_freq:
+                seq_freq[tokens] = 0
+            seq_freq[tokens] += word_frequencies[key]
+
+    return seq_freq
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or \
+            not isinstance(pair, tuple):
+        return None
+
+    preprocessed_freq = {}
+
+    for key, frequency in word_frequencies.items():
+        if str(pair)[1:-2] in str(key):
+            pair_indexes = []
+            for token_index in range(len(key) - 1):
+                tokens = (key[token_index], key[token_index + 1])
+                if tokens == pair:
+                    pair_indexes.append(token_index)
+
+            saved_tokens = list(key)
+            for index in pair_indexes:
```

Comment:
Congratulations on solving the secret! But there are some mistakes. Think about this line. It is in this place that you have problems and you get "ве" instead of "ле". Maybe debugging can be useful for you! Good luck!

Reply:
The problem is not "ве" and "ле", it's the shift "ел" -> "ев" -> "до". In the end, I just handled the exception in other part of code.